lib/engine/game/g_1873/step/buy_train.rb

```diff
@@ -345,6 +345,19 @@ def ebuy_president_can_contribute?(_corporation)
           def president_may_contribute?(_corporation, _active_shell)
             false
           end
+
+          def auto_actions(entity)
```

Comment:
these are all copy and pasted, maybe make a module and include it?

Reply:
I'm not sure exactly what you mean here. If it's about the auto_actions method, that's common to all classes that support programs. That seems like it should be addressed separately. If it's about the activate method, they're only identical in the first commit. The second commit changes them to read different program properties. Let me know if I've misunderstood.

Reply:
activate_program_independent_mines  is not correct? whta's the 'second commit' you mean you plan to have a different PR?

Reply:
activate_program_independent_mines is correct (AFAIK). The "second commit" is the second commit in the PR branch. In the combined change the methods are similar, but not identical.

Reply:
available_actions = actions(entity)
            return unless available_actions.include?('pass')
            return unless entity.minor?
            return unless program.skip_buy

            [Action::Pass.new(entity)]


this looks identical to me? what is different about them across the different steps?

Reply:
program.skip_buy is replaced by program.skip_close and program.skip_track in the other classes.

Reply:
got it thanks